Add new service package Support and new resource Case

mergenci opened the pull request:
<!---
See what makes a good Pull Request at: https://hashicorp.github.io/terraform-provider-aws/raising-a-pull-request/
--->
### Description
<!---
Please provide a helpful description of what change this pull request will introduce.
--->

Add new service package [Support](https://docs.aws.amazon.com/awssupport/latest/user/about-support-api.html) and new resource [Case](https://docs.aws.amazon.com/awssupport/latest/APIReference/API_CreateCase.html).

### Details

Terraform AWS Provider's [Contributor Guide](https://hashicorp.github.io/terraform-provider-aws/), [/docs directory](https://github.com/hashicorp/terraform-provider-aws/tree/main/docs) of the [repo](https://github.com/hashicorp/terraform-provider-aws), documents everything about provider development.

As part of this PR, we first [configured our development environment](https://hashicorp.github.io/terraform-provider-aws/development-environment/) by running `make tools` and then testing our setup by running `make build`.

In order to introduce our new service, AWS Support, we followed [Adding a New AWS Service](https://hashicorp.github.io/terraform-provider-aws/add-a-new-service/) section. More specifically, we applied the following steps.
1. [Edit](https://github.com/upbound/terraform-provider-aws/pull/132/files#diff-7a5da00a54a3e74a6be313e21473ed507e89c8c5df218f133d1c98b13a50f968) `names/names_data.csv`. Uncheck `NotImplemented` column, by removing “x”. Disable AWS SDK Go v1 client, by removing “1”, and enable AWS SDK Go v2 client, by adding “2” to the next column.
2. [Create](https://github.com/upbound/terraform-provider-aws/pull/132/files#diff-4048fe5fd184cc51bdb5dfa8f7dc1dfb603a20bc86f56ab3ca85b0e4f00f12e0) `internal/service/support/generate.go`.
3. Run `make gen` and `go mod tidy`.

Above commands generated AWS client ([`internal/conns/awsclient_gen.go`](https://github.com/upbound/terraform-provider-aws/pull/132/files#diff-99805755ac7d109b77122d2cccf89d22ef64c49b4fd1627ae67cba82b8e73019)) and configure the service package ([`internal/provider/service_packages_gen.go`](https://github.com/upbound/terraform-provider-aws/pull/132#diff-ff701934a38cdc4d6d650ebe0d3db3c90422ca68a0cc06b9f94991278c41ce2b)). 

Finally, we [implemented our new resource Case](https://github.com/upbound/terraform-provider-aws/pull/132/files#diff-a34d0170160712858490d4a0e2292139b9bdd626192378a3bb4a2e7353f15f63) by following [Adding a New Resource](https://hashicorp.github.io/terraform-provider-aws/add-a-new-resource/). Adding a resource [requires](https://hashicorp.github.io/terraform-provider-aws/add-a-new-resource/#register-resource-to-the-provider) another `make gen` that registers the resource to [`internal/service/support/service_package_gen.go`](https://github.com/upbound/terraform-provider-aws/pull/132/files#diff-d92afbf3e1030d1eb71c88cf8d33edb4990acd5aca9e56cb3bbdb3d6864e9ab1) We frequently referred to existing Plugin Framework resource implementations, which could be easily located by searching `@FrameworkResource` in the repository.

Contributor Guide has a dedicated [section for debugging](https://hashicorp.github.io/terraform-provider-aws/debugging/). First, we [step-debugged](https://hashicorp.github.io/terraform-provider-aws/debugging/#use-visual-studio-code-debugging) an [acceptance test](https://hashicorp.github.io/terraform-provider-aws/running-and-writing-acceptance-tests/), which creates resources in the cloud. Then, we end-to-end tested using Terraform CLI. In order to do so, we wrote a [.terraformrc](https://developer.hashicorp.com/terraform/cli/config/config-file), to point Terraform to our provider build with a [provider installation section](https://developer.hashicorp.com/terraform/cli/config/config-file#provider-installation).

mergenci committed:
Add new service package Support and new resource SupportCase.

Signed-off-by: Cem Mergenci <cmergenci@gmail.com>

ulucinar commented:
Hi @mergenci,
Thank you for introducing the new resource with the [Terraform plugin framework](https://developer.hashicorp.com/terraform/plugin/framework). Let's make sure that we also open an upstream PR for this resource. 

Could you please document the steps needed to add a new framework resource (including the generator commands). It would be helpful as a reference.

ulucinar commented on go.mod:
```diff
@@ -6,7 +6,7 @@ require (
 	github.com/YakDriver/regexache v0.23.0
 	github.com/aws/aws-sdk-go v1.49.2
-	github.com/aws/aws-sdk-go-v2 v1.24.0
+	github.com/aws/aws-sdk-go-v2 v1.24.1
```
Is this bump intended (something needed to implement the new resource)? If not, let's revert it back as it may negatively affect the other resources.

mergenci replied:
Version bump is performed by `go mod tidy`, because we use AWS Support SDK. AWS SDK Go v2 exposes each service in its own package. Support service [requires](https://github.com/aws/aws-sdk-go-v2/blob/service/support/v1.19.6/service/support/go.mod) AWS SDK Go v2 v1.24.1.

ulucinar commented on internal/service/support/support_case.go:
```diff
@@ -0,0 +1,261 @@
+	framework.ResourceWithConfigure
+}
+
+func (r *resourceSupportCase) Metadata(_ context.Context, request resource.MetadataRequest, response *resource.MetadataResponse) {
```
```suggestion
func (r *resourceSupportCase) Metadata(_ context.Context, _ resource.MetadataRequest, response *resource.MetadataResponse) {
```

mergenci replied:
Done.

ulucinar commented on internal/service/support/support_case.go:
```diff
@@ -0,0 +1,261 @@
+}
+
+func (r *resourceSupportCase) Metadata(_ context.Context, request resource.MetadataRequest, response *resource.MetadataResponse) {
+	response.TypeName = "aws_support_support_case"
```
How about naming the resource as `aws_support_case`?
I think we have chosen to name the resource using the convention `aws_<service name>_<resource name>`. I think Terraform resource names do not always follow this convention though, e.g., `aws_vpc`.

mergenci replied:
`names/data/names_data.csv` had ResourcePrefix already defined as `aws_support_`. In order to configure a different name, such as `aws_support_case`, we can write a new line.

mergenci replied:
[Contributor Guide suggests](https://hashicorp.github.io/terraform-provider-aws/naming/#rule_2) that we use “Case” as the resource name.

mergenci replied:
I renamed resource to “Case”.

ulucinar commented on internal/service/support/support_case.go:
```diff
@@ -0,0 +1,261 @@
+				Optional:    true,
+				ElementType: types.StringType,
+				Validators: []validator.List{
+					listvalidator.SizeAtMost(10),
```
Is this a limitation posed by the AWS API or the AWS Go SDK? A comment may help here.

mergenci replied:
This limit was [stated by AWS Support SDK documentation](https://docs.aws.amazon.com/awssupport/latest/APIReference/API_CreateCase.html#API_CreateCase_RequestSyntax). I tested whether eleven items would be accepted. They did. So I removed size validator.

ulucinar commented on internal/service/support/support_case.go:
```diff
@@ -0,0 +1,261 @@
+package support
+
+import (
+	"context"
+
+	"github.com/aws/aws-sdk-go-v2/service/support"
+	awstypes "github.com/aws/aws-sdk-go-v2/service/support/types"
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/hashicorp/terraform-plugin-framework-validators/listvalidator"
+	"github.com/hashicorp/terraform-plugin-framework-validators/stringvalidator"
+	"github.com/hashicorp/terraform-plugin-framework/path"
+	"github.com/hashicorp/terraform-plugin-framework/resource"
+	"github.com/hashicorp/terraform-plugin-framework/resource/schema"
+	"github.com/hashicorp/terraform-plugin-framework/resource/schema/listplanmodifier"
+	"github.com/hashicorp/terraform-plugin-framework/resource/schema/planmodifier"
+	"github.com/hashicorp/terraform-plugin-framework/resource/schema/stringplanmodifier"
+	"github.com/hashicorp/terraform-plugin-framework/schema/validator"
+	"github.com/hashicorp/terraform-plugin-framework/types"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/retry"
+	"github.com/hashicorp/terraform-provider-aws/internal/create"
+	"github.com/hashicorp/terraform-provider-aws/internal/errs"
+	"github.com/hashicorp/terraform-provider-aws/internal/framework"
+	"github.com/hashicorp/terraform-provider-aws/internal/framework/flex"
+	"github.com/hashicorp/terraform-provider-aws/internal/tfresource"
+	"github.com/hashicorp/terraform-provider-aws/names"
+)
```
I don't know the conventions in this repo but we generally group local imports (imports from `github.com/hashicorp/terraform-provider-aws`) together, separate from the other imports:
```suggestion
import (
	"context"

	"github.com/aws/aws-sdk-go-v2/service/support"
	awstypes "github.com/aws/aws-sdk-go-v2/service/support/types"
	"github.com/aws/aws-sdk-go/aws"
	"github.com/hashicorp/terraform-plugin-framework-validators/listvalidator"
	"github.com/hashicorp/terraform-plugin-framework-validators/stringvalidator"
	"github.com/hashicorp/terraform-plugin-framework/path"
	"github.com/hashicorp/terraform-plugin-framework/resource"
	"github.com/hashicorp/terraform-plugin-framework/resource/schema"
	"github.com/hashicorp/terraform-plugin-framework/resource/schema/listplanmodifier"
	"github.com/hashicorp/terraform-plugin-framework/resource/schema/planmodifier"
	"github.com/hashicorp/terraform-plugin-framework/resource/schema/stringplanmodifier"
	"github.com/hashicorp/terraform-plugin-framework/schema/validator"
	"github.com/hashicorp/terraform-plugin-framework/types"
	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/retry"

	"github.com/hashicorp/terraform-provider-aws/internal/create"
	"github.com/hashicorp/terraform-provider-aws/internal/errs"
	"github.com/hashicorp/terraform-provider-aws/internal/framework"
	"github.com/hashicorp/terraform-provider-aws/internal/framework/flex"
	"github.com/hashicorp/terraform-provider-aws/internal/tfresource"
	"github.com/hashicorp/terraform-provider-aws/names"
)
```

mergenci replied:
Done.

ulucinar commented on internal/service/support/support_case.go:
```diff
@@ -0,0 +1,261 @@
+}
+
+// Schema returns the schema for this resource.
+func (r *resourceSupportCase) Schema(ctx context.Context, request resource.SchemaRequest, response *resource.SchemaResponse) {
```
```suggestion
func (r *resourceSupportCase) Schema(_ context.Context, _ resource.SchemaRequest, response *resource.SchemaResponse) {
```

ulucinar commented on internal/service/support/support_case.go:
```diff
@@ -0,0 +1,261 @@
+}
+
+// Update is a no-op, for now. For another no-op update example, see service/auditmanager/organization_admin_account_registration.go
+func (r *resourceSupportCase) Update(ctx context.Context, request resource.UpdateRequest, response *resource.UpdateResponse) {
```
Does the `support.Client.AddCommunicationToCase` alter the existing communication parameters (communication body, CC email addresses, etc.) or does it add a new set as the name implies?

ulucinar commented on internal/service/support/support_case.go:
```diff
@@ -0,0 +1,261 @@
+}
+
+// Update is a no-op, for now. For another no-op update example, see service/auditmanager/organization_admin_account_registration.go
+func (r *resourceSupportCase) Update(ctx context.Context, request resource.UpdateRequest, response *resource.UpdateResponse) {
```
```suggestion
func (r *resourceSupportCase) Update(_ context.Context, _ resource.UpdateRequest, _ *resource.UpdateResponse) {
```

mergenci replied:
Done.

ulucinar commented on internal/service/support/support_case.go:
```diff
@@ -0,0 +1,261 @@
+func (r *resourceSupportCase) Update(ctx context.Context, request resource.UpdateRequest, response *resource.UpdateResponse) {
+}
+
+// Delete is a no-op, because AWS doesn't provide a deletion API. For another no-op update example, see service/auditmanager/organization_admin_account_registration.go
+func (r *resourceSupportCase) Delete(ctx context.Context, request resource.DeleteRequest, response *resource.DeleteResponse) {
```
```suggestion
// Delete is a no-op, because AWS doesn't provide a deletion API. For another no-op delete example, see service/auditmanager/organization_admin_account_registration.go
func (r *resourceSupportCase) Delete(_ context.Context, _ resource.DeleteRequest, _ *resource.DeleteResponse) {
```

mergenci replied:
Done.

ulucinar commented on internal/service/support/support_case.go:
```diff
@@ -0,0 +1,261 @@
+	// See internal/service/opensearchserverless/find.go findLifecyclePolicyByNameAndType()
+	if caseID == "" {
+		return nil, &retry.NotFoundError{
+			Message: "Cannot find SupportCase with an empty ID.",
```
```suggestion
			Message: "cannot find SupportCase with an empty ID.",
```

mergenci replied:
Done.

ulucinar commented on internal/service/support/support_case.go:
```diff
@@ -0,0 +1,261 @@
+		return nil, err
+	}
+
+	if output == nil || output.Cases == nil || len(output.Cases) == 0 {
```
```suggestion
	if output == nil || len(output.Cases) == 0 {
```

mergenci replied:
Done.

ulucinar commented on internal/service/support/support_case_test.go:
```diff
@@ -0,0 +1,108 @@
+package support_test
+
+import (
+	"context"
+	"errors"
+	"testing"
+
+	"github.com/aws/aws-sdk-go-v2/service/support"
+	"github.com/aws/aws-sdk-go/aws/endpoints"
+	"github.com/hashicorp/terraform-plugin-testing/helper/resource"
+	"github.com/hashicorp/terraform-plugin-testing/terraform"
+	"github.com/hashicorp/terraform-provider-aws/internal/acctest"
+	"github.com/hashicorp/terraform-provider-aws/internal/conns"
+	"github.com/hashicorp/terraform-provider-aws/internal/create"
+	tfsupport "github.com/hashicorp/terraform-provider-aws/internal/service/support"
+	"github.com/hashicorp/terraform-provider-aws/names"
+)
```
```suggestion
import (
	"context"
	"errors"
	"testing"

	"github.com/aws/aws-sdk-go-v2/service/support"
	"github.com/aws/aws-sdk-go/aws/endpoints"
	"github.com/hashicorp/terraform-plugin-testing/helper/resource"
	"github.com/hashicorp/terraform-plugin-testing/terraform"

	"github.com/hashicorp/terraform-provider-aws/internal/acctest"
	"github.com/hashicorp/terraform-provider-aws/internal/conns"
	"github.com/hashicorp/terraform-provider-aws/internal/create"
	tfsupport "github.com/hashicorp/terraform-provider-aws/internal/service/support"
	"github.com/hashicorp/terraform-provider-aws/names"
)
```

mergenci replied:
Done.

ulucinar reviewed (commented):
Thank you @mergenci for working on the new `SupportCase` resource. Left some minor comments. Let's discuss.

mergenci committed:
Remove documented-but-not-enforced size limit.

AWS Support SDK documentation states that ten CC email addresses are
allowed, but eleven email addresses are accepted in practice.

https://docs.aws.amazon.com/awssupport/latest/APIReference/API_CreateCase.html

Signed-off-by: Cem Mergenci <cmergenci@gmail.com>

mergenci committed:
Make code idiomatic and remove development comments.

Signed-off-by: Cem Mergenci <cmergenci@gmail.com>

mergenci committed:
Rename SupportCase to Case and add license headers.

According to resource naming rules in Contributor Guide, "Case" is a
better name than "SuportCase".

https://hashicorp.github.io/terraform-provider-aws/naming/#rule_2

Signed-off-by: Cem Mergenci <cmergenci@gmail.com>

mergenci committed:
Add license header.

Signed-off-by: Cem Mergenci <cmergenci@gmail.com>

ulucinar reviewed (approved):
Thanks @mergenci, lgtm. Thank you for implementing the new resource with the Terraform plugin framework.
